lib/src/gitignore.rs

```diff
@@ -14,189 +14,99 @@
-                        }
-                    }
-                }
-                if in_escape {
-                    regex.push_str(&regex_escape("\\"));
-                }
-                if i < components.len() - 1 {
-                    regex.push('/');
-                }
-            }
-        }
-        if matches_only_directory {
-            regex.push_str("/.*");
-        } else {
-            regex.push_str("(/.*|$)");
-        }
-        let regex = Regex::new(&regex).unwrap();
-
-        Some(GitIgnoreLine { is_negative, regex })
-    }
-
-    fn matches(&self, path: &str) -> bool {
-        self.regex.is_match(path)
-    }
+use std::{fs, io};
+
+use ignore::gitignore;
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum GitIgnoreError {
+    #[error("failed to read ignore patterns from file {path}: {source}")]
+    ReadFile { path: PathBuf, source: io::Error },
+    #[error("invalid UTF-8 for ignore pattern in {path} on line #{line_num_for_display} {line}")]
+    InvalidUtf8 {
+        path: PathBuf,
+        line_num_for_display: usize,
+        line: String,
+        source: std::str::Utf8Error,
+    },
+    #[error(transparent)]
+    Underlying(#[from] ignore::Error),
 }
 
 /// Models the effective contents of multiple .gitignore files.
 #[derive(Debug)]
 pub struct GitIgnoreFile {
-    parent: Option<Arc<GitIgnoreFile>>,
-    lines: Vec<GitIgnoreLine>,
+    path: String,
+    matchers: Vec<gitignore::Gitignore>,
 }
 
 impl GitIgnoreFile {
     pub fn empty() -> Arc<GitIgnoreFile> {
         Arc::new(GitIgnoreFile {
-            parent: None,
-            lines: vec![],
+            path: Default::default(),
+            matchers: Default::default(),
         })
     }
 
-    pub fn chain(self: &Arc<GitIgnoreFile>, prefix: &str, input: &[u8]) -> Arc<GitIgnoreFile> {
-        let mut lines = vec![];
-        for input_line in input.split(|b| *b == b'\n') {
-            // Skip non-utf8 lines
-            if let Ok(line_string) = String::from_utf8(input_line.to_vec()) {
-                if let Some(line) = GitIgnoreLine::parse(prefix, &line_string) {
-                    lines.push(line);
-                }
-            }
+    pub fn chain(
+        self: &Arc<GitIgnoreFile>,
+        prefix: &str,
+        input: &[u8],
+    ) -> Result<Arc<GitIgnoreFile>, GitIgnoreError> {
+        let path = self.path.clone() + prefix;
+        let mut builder = gitignore::GitignoreBuilder::new(&path);
+        for (i, input_line) in input.split(|b| *b == b'\n').enumerate() {
+            let line =
+                std::str::from_utf8(input_line).map_err(|err| GitIgnoreError::InvalidUtf8 {
+                    path: PathBuf::from(&path),
+                    line_num_for_display: i + 1,
+                    line: String::from_utf8_lossy(input_line).to_string(),
+                    source: err,
+                })?;
+            // TODO: do we need to provide the `from` argument? Is it for providing
```

Comment:
```suggestion
            // FIXME: do we need to provide the `from` argument? Is it for providing
```

If we don't plan to resolve this in this PR, let's change it to `FIXME`.